Allow SVG XML data to be passed in RML markup directly

mcukstorm opened the pull request:
At present the SVG element only supports loading SVG data from files, this change is to allow it to also be specified in the markup within the SVG tags. 

**Use case**
Implementing simple drawings into a page (could be a single path), or more likely being able to construct and specify SVG data via LUA for things like graph generation, in this scenario the SVG data would likely be loaded via innerRML / SetInnerRML

**Functional example usage (not a real world use case)**
```
<rml>
    <head></head>
    <body>
        <svg xmlns="http://www.w3.org/2000/svg" width="50" height="50" viewBox="0 0 50 50">
            <circle cx="25" cy="25" r="20" stroke="black" stroke-width="3" fill="red" />
        </svg>
    </body>
</rml>
```

**Implementation notes:** 
- As SVG XML doesn't have first party support in RmlUi it is treated as CDATA to prevent the XMLParser trying to handle it.
- The SVG XML is stored internally in an attribute(_cdata) on the element so it can be accessed later elsewhere.
- Uses the SVGCache not as it gives any benefits in this scenario but just to prevent duplicating SVG handling code elsewhere. 
- To facilitate using SVGCache the source that was used as the cache key and the xml source file path has changed to taking a source_id, source and source_type. For file based the source_id and source are the same.
- SVG elements not using src will be allocated an internal id stored in an attribute(_source-id) which is used as the source_id for SVGCache.

**Wider implications:** Allows plugins to register elements as containing CDATA rather than PCDATA, this does cause some incompatibilities with using SetInnerRML on such elements, in this case this method is overridden in the ElementSVG to have a different implementation. I toyed with changing the Factory methods (InstanceElementText/InstanceElementStream) to cause content changes of an element registered as containing CDATA to be re-processed by the registered XMLNodeHandler but in this situation the XMLParser state had to many differences to what it would be when doing the initial xml parsing. (If you wrap the CDATA in the expected parent tag, it re-creates an element, if you don't the internal check to see if it should be CDATA doesn't happen)

**Final Note** I'm also not an experienced c++ developer and also appreciate how some of these are implemented are probably less than ideal and happy to adapt to better fit with your views and best practices.

mcukstorm commented:
The most recent change in this PR should overcome the passing of the CDATA to the relevant XMLNodeHandler rather than overriding the SetInnerRML function for the element as was done initially.

mikke89 commented:
This is great, thanks a lot!

I'll have to comb through the pull request a bit later on, as I'll be traveling soon. But this is a very welcome feature, and has been requested multiple times earlier. For reference, here are a couple of related issues/PRs: #238 #591.

mcukstorm commented:
Thanks for your prompt comment, there is no rush on review. 

I've just added another change which removes the usage of the custom attributes to store the cdata/source id and moved this to a text node child, this is inserted as a non_dom to prevent rendering, though upon writing this I've realised this will prevent GetInnerRML working so may revise this. This structure fits better conceptually and feels a bit cleaner.

mcukstorm commented:
@mikke89 just a gentle reminder about this PR, would you be able to take a look at some point. I'm still happy to make any changes required.

mikke89 commented:
Yeah, this has taken way too long on my part, sorry about that! Thanks for reminding me, I'm putting this one on the top of my list now.

mikke89 commented on Source/SVG/XMLNodeHandlerSVG.h:
```diff
@@ -0,0 +1,65 @@
+#include <random>
+
+#ifndef RMLUI_CORE_XMLNODEHANDLERSVG_H
+	#define RMLUI_CORE_XMLNODEHANDLERSVG_H
```
Header guards should go above includes. And it also seems to include a lot more files than necessary.

mikke89 commented on Source/SVG/XMLNodeHandlerSVG.cpp:
```diff
@@ -0,0 +1,75 @@
+		// Only handle CDATA (tag check shouldn't really be necessary)
+		if (tag == "svg" && type == XMLDataType::CDATA)
+		{
+			auto* parent = rmlui_static_cast<ElementSVG*>(parser->GetParseFrame()->element);
```
I think we should use `rmlui_dynamic_cast` and verify the result for some added safety guards.

mikke89 commented on Source/SVG/XMLNodeHandlerSVG.cpp:
```diff
@@ -0,0 +1,75 @@
+			auto* parent = rmlui_static_cast<ElementSVG*>(parser->GetParseFrame()->element);
+			RMLUI_ASSERT(parent);
+
+			// Try to get the existing text node, or create if its missing (it is inserted as a non DOM child so should be at the start)
```
I don't think we really need to use an ElementText element at all here. I think we might as well just add the CData directly to the parent SVG element, and let that one handle the data directly.

That way, most of the code here can be removed or moved into the SVG element.

mikke89 commented on Tests/Data/VisualTests/svg_inline.rml:
```diff
@@ -0,0 +1,10 @@
+		<title>Top Bar</title>
+	</head>
+	<body>
+	    <svg xmlns="http://www.w3.org/2000/svg" width="100" height="100">
```
I like the intention here of adding this. But it doesn't really give much more than the sample itself, and it also forces everyone to use the SVG plugins when running the tests. I think leaving this to the sample only for now would be preferred.

mikke89 commented on Source/SVG/ElementSVG.cpp:
```diff
@@ -110,23 +103,89 @@ void ElementSVG::OnPropertyChange(const PropertyIdSet& changed_properties)
+				continue;
+			svg_data.append(attr.first);
+			svg_data.append("=\"");
+			svg_data.append(StringUtilities::Replace(attr.second.Get<String>(), "\"", "&quot;"));
```
Can use `StringUtilities::EncodeRml` here.

mikke89 commented on Source/Core/Factory.cpp:
```diff
@@ -354,6 +354,12 @@ ElementPtr Factory::InstanceElement(Element* parent, const String& instancer_nam
+	{
+		parser.HandleData(in_text, XMLDataType::CDATA);
+		return true;
+	}
```
This change worries me a bit. It doesn't seem right for this function to branch based on the parent, and even knowing about cdata doesn't seem right here.

I wonder if the more appropriate thing is to overload the `SetInnerRml` call? Then we should also get away with removing most modifications to the XML parser (except the CData setter). We should measure that making that function virtual is not causing any significant performance degradation, but I don't suspect as much.

I'm still open about this one, but at least I think we can make some improvements to better isolate responsibilities. An alternative is perhaps to set the `parse_as_rml` to true, and then I think it should essentially do the same thing.

Any thoughts?

mcukstorm commented:
Thank you for taking the time to review it and your feedback. I'm sorry its taken so long to reply. I've tried a couple of approaches and think your suggestion of overriding SetInnerRML is best.

I've attached 2 benchmark runs for master & this PR branch, the Table SetInnerRML benchmark seems to show the biggest difference.
[master-benchmark-run1.txt](https://github.com/user-attachments/files/23293922/master-benchmark-run1.txt)
[svg-inline-benchmark-run1.txt](https://github.com/user-attachments/files/23293921/svg-inline-benchmark-run1.txt)
[master-benchmark-run2.txt](https://github.com/user-attachments/files/23293923/master-benchmark-run2.txt)
[svg-inline-benchmark-run2.txt](https://github.com/user-attachments/files/23293924/svg-inline-benchmark-run2.txt)

I've not been able to reproduce the issues with DRMLUI_PRECOMPILED_HEADERS=OFF

The issue with tiger.svg loading is now fixed - it was using the provided src rather than the path including the document source path.

Generally the changes made are as you suggested, points of note:-
- BaseXMLParser.cpp - Moved the ToLower inside the IsCDATATag check so at least it only doubles up on ToLower case for tags marked as CDATA.
- I've updated the SVG sample to include examples of using data-rml with the SVG element.
